[i18n] Added translation feature for label and description of channels inside a channel-group-type

cweitkamp opened the pull request:
- Added translation feature for label and description of channels inside a `channel-group-type`
- Refactoring for centralized handling of translations in `ThingTypeI18nLocalizationService`
- Updated documentation

See https://github.com/openhab/openhab2-addons/pull/3678#issuecomment-397965219

Signed-off-by: Christoph Weitkamp <github@christophweitkamp.de>

cweitkamp commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java:
```diff
@@ -65,29 +72,67 @@ protected void unsetChannelTypeRegistry(ChannelTypeRegistry channelTypeRegistry)
+        String description = thingTypeI18nUtil.getChannelGroupDescription(bundle, channelGroupTypeUID,
+                channelGroupType.getDescription(), locale);
+
+        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(
```
The lines 119-142 are almost identical to 154-176 but I could not find a way to extract a method for them. Any thoughts?

maggu2810 replied:
Difficult to identify the problem in the Github diff view only. What's the problem?

cweitkamp replied:
@maggu2810 Thanks for your review.

This is the duplicate code which I want to extract into a new method. The problem is that I want to determine the `channelLabel` and `channelDescription` inside the foreach loop derived from either the `thingTypeUID` or the `channelGroupTypeUID` - I marked the code with **.

```java
    private List<ChannelDefinition> createLocalizedChannelDefinitions(Bundle bundle,
            List<ChannelDefinition> channelDefinitions, Locale locale) {
        // start
        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(channelDefinitions.size());
        for (ChannelDefinition channelDefinition : channelDefinitions) {
            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, **<thingTypeUID|channelGroupTypeUID>**, channelDefinition,
                    channelDefinition.getLabel(), locale);
            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, **<thingTypeUID|channelGroupTypeUID>**, channelDefinition,
                    channelDefinition.getDescription(), locale);
            if (channelLabel == null || channelDescription == null) {
                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
                if (channelType != null) {
                    if (channelLabel == null) {
                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
                                locale);
                    }
                    if (channelDescription == null) {
                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
                                channelType.getDescription(), locale);
                    }
                }
            }
            localizedChannelDefinitions
                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
                            channelDefinition.getProperties(), channelLabel, channelDescription));
        }
        // end
        return localizedChannelDefinitions;
    }
```

maggu2810 replied:
```diff
diff --git a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
index df5c1c14fb..2a2b50c766 100644
--- a/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
+++ b/bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/i18n/ThingTypeI18nLocalizationService.java
@@ -15,6 +15,7 @@ package org.eclipse.smarthome.core.thing.i18n;
 import java.util.ArrayList;
 import java.util.List;
 import java.util.Locale;
+import java.util.function.Function;
 
 import org.eclipse.jdt.annotation.NonNullByDefault;
 import org.eclipse.jdt.annotation.Nullable;
@@ -116,31 +117,12 @@ public class ThingTypeI18nLocalizationService {
         String description = thingTypeI18nUtil.getChannelGroupDescription(bundle, channelGroupTypeUID,
                 channelGroupType.getDescription(), locale);
 
-        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(
-                channelGroupType.getChannelDefinitions().size());
-        for (ChannelDefinition channelDefinition : channelGroupType.getChannelDefinitions()) {
-            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelGroupTypeUID, channelDefinition,
-                    channelDefinition.getLabel(), locale);
-            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelGroupTypeUID,
-                    channelDefinition, channelDefinition.getDescription(), locale);
-            if (channelLabel == null || channelDescription == null) {
-                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
-                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
-                if (channelType != null) {
-                    if (channelLabel == null) {
-                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
-                                locale);
-                    }
-                    if (channelDescription == null) {
-                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
-                                channelType.getDescription(), locale);
-                    }
-                }
-            }
-            localizedChannelDefinitions
-                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
-                            channelDefinition.getProperties(), channelLabel, channelDescription));
-        }
+        List<ChannelDefinition> localizedChannelDefinitions = localizeChannelDefinitions(bundle, channelGroupType.getChannelDefinitions(),
+                channelDefinition -> thingTypeI18nUtil.getChannelLabel(bundle, channelGroupTypeUID, channelDefinition,
+                        channelDefinition.getLabel(), locale),
+                channelDefinition -> thingTypeI18nUtil.getChannelDescription(bundle, channelGroupTypeUID,
+                        channelDefinition, channelDefinition.getDescription(), locale),
+                locale);
 
         return new ChannelGroupType(channelGroupTypeUID, channelGroupType.isAdvanced(), label, description,
                 channelGroupType.getCategory(), localizedChannelDefinitions);
@@ -151,30 +133,12 @@ public class ThingTypeI18nLocalizationService {
         String label = thingTypeI18nUtil.getLabel(bundle, thingTypeUID, thingType.getLabel(), locale);
         String description = thingTypeI18nUtil.getDescription(bundle, thingTypeUID, thingType.getDescription(), locale);
 
-        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(thingType.getChannelDefinitions().size());
-        for (ChannelDefinition channelDefinition : thingType.getChannelDefinitions()) {
-            String channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, thingTypeUID, channelDefinition,
-                    channelDefinition.getLabel(), locale);
-            String channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, thingTypeUID, channelDefinition,
-                    channelDefinition.getDescription(), locale);
-            if (channelLabel == null || channelDescription == null) {
-                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
-                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
-                if (channelType != null) {
-                    if (channelLabel == null) {
-                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
-                                locale);
-                    }
-                    if (channelDescription == null) {
-                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
-                                channelType.getDescription(), locale);
-                    }
-                }
-            }
-            localizedChannelDefinitions
-                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
-                            channelDefinition.getProperties(), channelLabel, channelDescription));
-        }
+        List<ChannelDefinition> localizedChannelDefinitions = localizeChannelDefinitions(bundle, thingType.getChannelDefinitions(),
+                channelDefinition -> thingTypeI18nUtil.getChannelLabel(bundle, thingTypeUID, channelDefinition,
+                        channelDefinition.getLabel(), locale),
+                channelDefinition -> thingTypeI18nUtil.getChannelDescription(bundle, thingTypeUID, channelDefinition,
+                        channelDefinition.getDescription(), locale),
+                locale);
 
         List<ChannelGroupDefinition> localizedChannelGroupDefinitions = new ArrayList<>(
                 thingType.getChannelGroupDefinitions().size());
@@ -213,4 +177,32 @@ public class ThingTypeI18nLocalizationService {
         return builder.build();
     }
 
+    public <T> List<ChannelDefinition> localizeChannelDefinitions(final Bundle bundle, final List<ChannelDefinition> channelDefinitions,
+            final Function<ChannelDefinition, String> channelLabelResolver,
+            final Function<ChannelDefinition, String> channelDescriptionResolver, final Locale locale) {
+        List<ChannelDefinition> localizedChannelDefinitions = new ArrayList<>(channelDefinitions.size());
+        for (ChannelDefinition channelDefinition : channelDefinitions) {
+            String channelLabel = channelLabelResolver.apply(channelDefinition);
+            String channelDescription = channelDescriptionResolver.apply(channelDefinition);
+            if (channelLabel == null || channelDescription == null) {
+                ChannelTypeUID channelTypeUID = channelDefinition.getChannelTypeUID();
+                ChannelType channelType = channelTypeRegistry.getChannelType(channelTypeUID, locale);
+                if (channelType != null) {
+                    if (channelLabel == null) {
+                        channelLabel = thingTypeI18nUtil.getChannelLabel(bundle, channelTypeUID, channelType.getLabel(),
+                                locale);
+                    }
+                    if (channelDescription == null) {
+                        channelDescription = thingTypeI18nUtil.getChannelDescription(bundle, channelTypeUID,
+                                channelType.getDescription(), locale);
+                    }
+                }
+            }
+            localizedChannelDefinitions
+                    .add(new ChannelDefinition(channelDefinition.getId(), channelDefinition.getChannelTypeUID(),
+                            channelDefinition.getProperties(), channelLabel, channelDescription));
+        }
+        return localizedChannelDefinitions;
+    }
+
 }
```

cweitkamp replied:
Very nice. Works like a charm. Thank you very much.

maggu2810 replied:
Now I will leave it to another committer to press the merge button :wink:

lolodomo commented:
Just for my understanding: what is wrong exactly ? I am running the openHAB snapshot 1302 and I get the correct translations in French in Paper UI with the WU binding.

cweitkamp commented:
@lolodomo The weather binding does not contain an example for this. Try to translate the label and description in 

https://github.com/cweitkamp/smarthome/blob/edec55aa1f0833db84e2e056b18ca862fb49c529/bundles/core/org.eclipse.smarthome.core.thing.xml.test/src/test/resources/test-bundle-pool/yahooweather.bundle/ESH-INF/thing/thing-types.xml#L43-46

cweitkamp commented:
@maggu2810 Sry to break your approval. I had to rebase because of #5864. Thus I used the opportunity to switch to the new ChannelTypeBuilder API. I additionally introduced a new ChannelGroupTypeBuilder API.

maggu2810 commented:
@cweitkamp I agree with the `ChannelGroupTypeBuilder`. If the code of the i18n stuff has been changed only (because of the nullness annotations), I assume my review is still valid.

maggu2810 commented:
@triller-telekom WRT to the builder, would you like to have a look at, too?

cweitkamp commented:
> If the code of the i18n stuff has been changed only (because of the nullness annotations)

I have added annotations in my i18n part and switched to the new `I18nUtil.stripConstantOr()` method.

triller-telekom commented on bundles/core/org.eclipse.smarthome.core.thing/src/main/java/org/eclipse/smarthome/core/thing/type/ChannelGroupTypeBuilder.java:
```diff
@@ -0,0 +1,114 @@
+    /**
+     * Specify whether this is an advanced {@link ChannelGroupType}, default is false
+     *
+     * @param advanced true is this is an advanced {@link ChannelGroupType}
```
typo: "if" this is an...

triller-telekom commented:
@maggu2810 The typo I found is fixed now, feel free to merge if travis agrees :)

triller-telekom commented:
Upps... Thanks, fixed now with https://github.com/eclipse/smarthome/pull/5883 :)

cweitkamp commented:
Thank you, too.

cweitkamp commented:
@maggu2810 Thanks
